README.md
I had trouble finding a copy of this (http://ftp.gwdg.de/pub/linux/handhelds/feeds/bootldr/bootldr-2.21.12.tar.gz), so I figured I'd put it up on github.

NOTE: I compiled it but haven't tested it, yet...

To build it, I downloaded and unpacked these into /usr/local/arm:

- http://ftp.gwdg.de/pub/linux/handhelds/projects/toolchain/archive/arm-linux-libz-2.95.tar.gz
- http://ftp.gwdg.de/pub/linux/handhelds/projects/toolchain/archive/cross-2.95.3.tar.bz2



CHANGELOG
LATEST STABLE BOOTLDR IS 2.18.54
LATEST EXTREMELY STABLE BOOTLDR IS 2.18.01

2.21.12: jamey 03-nov-03
  -- built bootldr with arm-linux-gcc 2.95.3 for sa1110 ipaqs
2.21.10:
2003-09-24 18:33  tag BOOTLDR_2_21_10
2003-09-24 18:33  jamey
	* Makefile: 2.21.10
2003-09-24 18:30  nelson
	* vfat.c: Dang!  I thought I'd checked this fix in.  This fixes the
	  "file not found" error, or at least it did some months ago.
2003-09-24 12:46  bavery
	* Makefile, params.h, pcmcia.c: fixed the linuxargs to be 2.4.19
	  compliant, reduced some verbosity
2003-09-23 11:37  bavery
	* Makefile, bootldr.c, btflash.c, ide.c, params.h, pcmcia.c,
	  pcmcia.h, serial.c: aded skiff fixes for timer,pcmcia and ide.
2003-09-23 11:31  bavery
	* skiff.h:  skiff specific stuff
2003-09-17 15:46  bavery
	* load_kernel/src/Makefile:  also work with config.local.mk
2003-09-17 15:46  jamey
	* Makefile: define SHORT_ARCH_STR for skiff
2.21.2: 2003-09-17 13:35  bavery
	* load_kernel/src/: Makefile, jffs2.c: skiff updates, allowed arch
	  flags to reach here and use them to ifdef out lcd spec stuff.
	* serial.c:  skiff updates - made structures for skiff
	* sa1100.h:  skiff updates
	* params.h: skiff updates
	* os_linux.mk: removed LINUX_DIR as it is unused and confuses gcc
	  3.2.3 also, added an auto detect to where zlib is so it works off
	  of the existing toolchain. i.e. it looks in the base dir of
	  arm-linux-gcc /arm-linux/{include,lib} for the zlib stuff
	* mmu-pxa.c: added PXA: to mmu message so we know which is which
	* flashif.c: skiff updates
	* config.mk: skiff updates, added some info to help know what
	  setting to use
	* btpci.c: skiff updates
	* btflash.c: skiff updates esp make atmel CB flash sectors right
	* bootlinux.c:	skiff updates includeing the dc21285: for mtdparts
	* bootldr.c, bootldr.h: skiff updates
	* boot.s:  skif updates, added assembeler based serial print for
	  debugging and added boot_flag space to teh stack
	* Makefile, boot_flags.h, bootcmds.c, bootconfig.h: skiff updates
2.20.5:
2003-09-17 13:04  bavery
	* Makefile: made a bogus miro version 5 to provide a tag point
	  before the large skiff checkin

2003-09-06 15:00  joshua
	* serial.c: RXSTAT error stuff removed. If you want it back,
	  compile with -DCONFIG_CONFUSE_USERS.

2003-08-25 14:11  joshua
	* bootldr.c: memtest2 and prelim 1900

2003-08-25 14:09  joshua
	* Makefile: -DCONFIG_USB, not CONFIG_USB.

2003-07-23 07:52  jamey
	* bootldr.c: updated help strings to indicate receive by ymodem or
	  xmodem

2003-07-08 08:50  jamey
	* architecture.h: added XDA_MACH

2003-07-03 10:54  pb
	* boot-pxa.s: change ldrh to ldr to avoid addressing range problems

2003-07-03 10:45  jamey
	* Makefile: updated makefile to include -sa or -pxa in bootldr
	  binary file names

2.20.4: 2003-07-03 10:40  jamey
	* CHANGELOG, Makefile, boot-sa1100.s: roll back to boot-sa1100.s to
	  avoid making bricks

2003-06-28 14:20  joshua
	* boot-pxa.s: boot-pxa update for h1910
	* boot-pxa.s: we can now check action button on 1910

2003-06-18 21:40  joshua
	* boot-pxa.s: Checked in boot-pxa.s which can detect h1900.
	  ***WARNING WARNING!*** THIS CODE MAY NOT WORK AND MAY FAIL TO
	  BOOT! Please test this with ram-bootldr BEFORE flashing to your
	  device!
	* boot_flags.h, bootldr.c, bootldr.h: added defines for h1900 and
	  axim in headers. now to implement..
	* Makefile: fixed USB so that enabling it doesn't break PXA builds.
	* asm-arm/arch-pxa/: h1900-gpio.h, h3900-irqs.h: added gpios for
	  h1900
	* asm-arm/arch-pxa/: axim-gpio.h, h3900-gpio.h: added axim-gpio.h
	* asm-arm/arch-pxa/pxa-regs.h: oops. stupid __REG.
	* asm-arm/arch-pxa/pxa-regs.h: synched pxa-regs.h up to kernel.
	* dma-sa1100.c, usb-sa1110.c: it seems I have changes that I never
	  committed.

2.20.3: jamey 03-JUL-03
  -- rollback boot-sa1100.s to the version from 2.19.73 because it was
  bricking some h3600's.  Will try to come up with a better solution
  when I have one of the bricks to debug with.
2.20.2: pb 06-JUN-03
  -- support action button on h5400
2.20.1: jamey 13-MAY-03
  -- update to 2.20.x to declare stability
  -- major new bootldr hacking will be in linux as bootldr
USE AT YOUR OWN RISK!!!!
2.19.76: pb 29-APR-03
  -- use ATAG_INITRD2 for initrd so that addresses work out right on pxa.
2.19.75: pb 25-APR-03
  -- added extra (optional) machine-type column to reflash.ctl.
  -- tolerate reflash.ctl not ending with newline.
  -- correct left/right joypad confusion on h3900.
  -- fix pxa cache-flushing code.  enable both I & D caches as standard on h3900.
  -- mark RAM non-bufferable on pxa in deference to writeback cache errata.
2.19.74: jamey 23-APR-03
  -- changed boot-sa1100.s to use same SDRAM config values (MDCNFG and MDREFR) as Linux
  -- applied patch from disconnect for initial reflash save support' reflash.c
  -- reduced verbosity of flash_iohandle_prewrite
  -- allow tabs as well as spaces in reflash.ctl
2.19.73: joshua 9-APR-03
  -- usb send over DMA works for real. sorry.
2.19.72: joshua 5-APR-03
  -- jamey's cut-n-paste-happyness in CHANGELOG fixed
  -- usb send support :)
2.19.71: jamey 25-MAR_03
 -- dual bank probe_ram works on h3900 with dual sdram banks, but
    DABT's on single bank h3900.
2.19.70: joshua 15-MAR-03
  -- experimental USB support added.
2.19.69: jamey 14-MAR-03
  -- build with -march=armv4 so that flashif.c will work with 16bit
     flash
  -- check for armv4 option
  -- set flashword = UNCACHED_FLASH_BASE in ram bootldr
2.19.68: jamey 14-MAR-03
  -- debugged support for reflashing bootldr, hopefully it is pretty safe.
  -- enable ymodem by default
2.19.67: jamey 13-MAR-03
  -- add support for reflashing bootldr, including validating bootldr
     image for architecture
2.19.66: pb 7-MAR-03
  -- reflash can handle multiple entries in one md5sum file.
2.19.65: jamey 6-MAR-03
  -- LCD working again on h3900 but h5400 buzzing incessantly
  -- verified action button detection working on h3900
  -- Joshua Wise added a label for reflash to the splash screen.

md5.c
/* MD5 support for bootldr
 * ^^^^^^^^^^^^^^^^^^^^^^^
 * Copyright (C) 2000, 2001  John G Dorsey
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * The author may be contacted via electronic mail at <john+@cs.cmu.edu>,
 * or at the following address:
 *
 *   John Dorsey
 *   Carnegie Mellon University
 *   HbH2201 - ICES
 *   5000 Forbes Avenue
 *   Pittsburgh, PA  15213
 *
 *
 * History:
 * ^^^^^^^^
 * 12 March, 2001 - created by separating from jffs.c. (jd)
 *
 */

#ifndef CONFIG_ACCEPT_GPL
#error This file covered by GPL but CONFIG_ACCEPT_GPL undefined.
#endif

#include <string.h>

#include "md5.h"

/* MD5 Message Digest Algorithm - IETF Request for Comments: 1321
 *
 * We provide a (hopefully) conforming implementation here to verify
 * the correctness of reconstructed JFFS files and downloaded kernel
 * images.
 */

/* Auxiliary function "F" from RFC 1321 section 3.4: */
static inline unsigned int md5_F(unsigned int X, unsigned int Y,
				 unsigned int Z){
  return (X & Y) + (~X & Z);
}

/* Auxiliary function "G" from RFC 1321 section 3.4: */
static inline unsigned int md5_G(unsigned int X, unsigned int Y,
				 unsigned int Z){
  return (X & Z) + (Y & ~Z);
}

/* Auxiliary function "H" from RFC 1321 section 3.4: */
static inline unsigned int md5_H(unsigned int X, unsigned int Y,
				 unsigned int Z){
  return X ^ Y ^ Z;
}

/* Auxiliary function "I" from RFC 1321 section 3.4: */
static inline unsigned int md5_I(unsigned int X, unsigned int Y,
				 unsigned int Z){
  return Y ^ (X | ~Z);
}

/* Rotate ("<<<") operation from RFC 1321 section 3.4: */
static inline unsigned int md5_rotate(unsigned int a, int r){
  return (a << r) | (a >> ((sizeof(a) * 8) - r));
}

/* Compute the number of padding and append-length bytes necessary 
 * for the given `length'. (See RFC 1321 sections 3.1, 3.2)
 */
unsigned int md5_extend(unsigned int length){
  unsigned int ext;
  
  ext = (MD5_LENGTH_MODULUS - (length % MD5_LENGTH_MODULUS)) % 
    MD5_LENGTH_MODULUS;

  if(ext < MD5_PREPAD_LEN_LENGTH)
    ext += MD5_LENGTH_MODULUS;
  
  return ext;
}


/* Run the MD5 algorithm over each block in the file. (Sections 3.3 and
 * 3.4 of the RFC.)
 */
static void md5_rounds(unsigned char *data, unsigned int length,
		       unsigned int *buf){

  unsigned int A, B, C, D, AA, BB, CC, DD, X[MD5_BLOCK_WORDS];
  unsigned int T[] = MD5_T_INIT;
  unsigned int *M = (unsigned int *)data;
  unsigned int words;
  int i, j;

  A = MD5_WORD_A_INIT;
  B = MD5_WORD_B_INIT;
  C = MD5_WORD_C_INIT;
  D = MD5_WORD_D_INIT;
  
  words = (length + md5_extend(length)) / sizeof(unsigned int);

  for(i = 0; i < (words / MD5_BLOCK_WORDS); ++i){

    for(j = 0; j < MD5_BLOCK_WORDS; ++j)
      X[j] = M[(i * MD5_BLOCK_WORDS) + j];

    AA = A;
    BB = B;
    CC = C;
    DD = D;

#define MD5_ROUND1(a, b, c, d, k, s, i) \
((a) = (b) + md5_rotate((a) + \
 md5_F((b), (c), (d)) + X[(k)] + T[(i) - 1], (s)))

    MD5_ROUND1(A, B, C, D,  0,  7,  1);
    MD5_ROUND1(D, A, B, C,  1, 12,  2);
    MD5_ROUND1(C, D, A, B,  2, 17,  3);
    MD5_ROUND1(B, C, D, A,  3, 22,  4);

    MD5_ROUND1(A, B, C, D,  4,  7,  5);
    MD5_ROUND1(D, A, B, C,  5, 12,  6);
    MD5_ROUND1(C, D, A, B,  6, 17,  7);
    MD5_ROUND1(B, C, D, A,  7, 22,  8);

    MD5_ROUND1(A, B, C, D,  8,  7,  9);
    MD5_ROUND1(D, A, B, C,  9, 12, 10);
    MD5_ROUND1(C, D, A, B, 10, 17, 11);
    MD5_ROUND1(B, C, D, A, 11, 22, 12);

    MD5_ROUND1(A, B, C, D, 12,  7, 13);
    MD5_ROUND1(D, A, B, C, 13, 12, 14);
    MD5_ROUND1(C, D, A, B, 14, 17, 15);
    MD5_ROUND1(B, C, D, A, 15, 22, 16);

#define MD5_ROUND2(a, b, c, d, k, s, i) \
((a) = (b) + md5_rotate((a) + \
 md5_G((b), (c), (d)) + X[(k)] + T[(i) - 1], (s)))

    MD5_ROUND2(A, B, C, D,  1,  5, 17);
    MD5_ROUND2(D, A, B, C,  6,  9, 18);
    MD5_ROUND2(C, D, A, B, 11, 14, 19);
    MD5_ROUND2(B, C, D, A,  0, 20, 20);

    MD5_ROUND2(A, B, C, D,  5,  5, 21);
    MD5_ROUND2(D, A, B, C, 10,  9, 22);
    MD5_ROUND2(C, D, A, B, 15, 14, 23);
    MD5_ROUND2(B, C, D, A,  4, 20, 24);

    MD5_ROUND2(A, B, C, D,  9,  5, 25);
    MD5_ROUND2(D, A, B, C, 14,  9, 26);
    MD5_ROUND2(C, D, A, B,  3, 14, 27);
    MD5_ROUND2(B, C, D, A,  8, 20, 28);

    MD5_ROUND2(A, B, C, D, 13,  5, 29);
    MD5_ROUND2(D, A, B, C,  2,  9, 30);
    MD5_ROUND2(C, D, A, B,  7, 14, 31);
    MD5_ROUND2(B, C, D, A, 12, 20, 32);

#define MD5_ROUND3(a, b, c, d, k, s, i) \
((a) = (b) + md5_rotate((a) + \
 md5_H((b), (c), (d)) + X[(k)] + T[(i) - 1], (s)))

    MD5_ROUND3(A, B, C, D,  5,  4, 33);
    MD5_ROUND3(D, A, B, C,  8, 11, 34);
    MD5_ROUND3(C, D, A, B, 11, 16, 35);
    MD5_ROUND3(B, C, D, A, 14, 23, 36);

    MD5_ROUND3(A, B, C, D,  1,  4, 37);
    MD5_ROUND3(D, A, B, C,  4, 11, 38);
    MD5_ROUND3(C, D, A, B,  7, 16, 39);
    MD5_ROUND3(B, C, D, A, 10, 23, 40);

    MD5_ROUND3(A, B, C, D, 13,  4, 41);
    MD5_ROUND3(D, A, B, C,  0, 11, 42);
    MD5_ROUND3(C, D, A, B,  3, 16, 43);
    MD5_ROUND3(B, C, D, A,  6, 23, 44);

    MD5_ROUND3(A, B, C, D,  9,  4, 45);
    MD5_ROUND3(D, A, B, C, 12, 11, 46);
    MD5_ROUND3(C, D, A, B, 15, 16, 47);
    MD5_ROUND3(B, C, D, A,  2, 23, 48);

#define MD5_ROUND4(a, b, c, d, k, s, i) \
((a) = (b) + md5_rotate((a) + \
 md5_I((b), (c), (d)) + X[(k)] + T[(i) - 1], (s)))

    MD5_ROUND4(A, B, C, D,  0,  6, 49);
    MD5_ROUND4(D, A, B, C,  7, 10, 50);
    MD5_ROUND4(C, D, A, B, 14, 15, 51);
    MD5_ROUND4(B, C, D, A,  5, 21, 52);

    MD5_ROUND4(A, B, C, D, 12,  6, 53);
    MD5_ROUND4(D, A, B, C,  3, 10, 54);
    MD5_ROUND4(C, D, A, B, 10, 15, 55);
    MD5_ROUND4(B, C, D, A,  1, 21, 56);

    MD5_ROUND4(A, B, C, D,  8,  6, 57);
    MD5_ROUND4(D, A, B, C, 15, 10, 58);
    MD5_ROUND4(C, D, A, B,  6, 15, 59);
    MD5_ROUND4(B, C, D, A, 13, 21, 60);

    MD5_ROUND4(A, B, C, D,  4,  6, 61);
    MD5_ROUND4(D, A, B, C, 11, 10, 62);
    MD5_ROUND4(C, D, A, B,  2, 15, 63);
    MD5_ROUND4(B, C, D, A,  9, 21, 64);

    A += AA;
    B += BB;
    C += CC;
    D += DD;

  }

  buf[0] = A;
  buf[1] = B;
  buf[2] = C;
  buf[3] = D;
}


/* Initialize the padded postamble to the reconstructed file. (Sections
 * 3.1 and 3.2 of the RFC.)
 */
static void md5_prep(unsigned char *data, unsigned int length){
  unsigned int extend;

  extend = md5_extend(length);
  
  memset(data + length, 0, extend);

  data[length] = 0x80;

  *((unsigned int *)(data + length + 
		     extend - (2 * sizeof(unsigned int)))) = length << 3;
  *((unsigned int *)(data + length + 
		     extend - (1 * sizeof(unsigned int)))) = length >> 29;
}


/* Modify the postamble section of the buffer pointed to by `data' having
 * `length' bytes (with appropriate padding sized by md5_extend()), then
 * return the MD5 sum in a four-integer array pointed to by `sum':
 */
void md5_sum(unsigned char *data, unsigned int length, unsigned int *sum){

  md5_prep(data, length);

  md5_rounds(data, length, sum);

}

/* Display the result of an MD5 computation (section 3.5 of the RFC): */
void md5_display(unsigned int *sum){
  int i, j;

  for(i = 0; i < MD5_SUM_WORDS; ++i)
    for(j = 0; j < sizeof(unsigned int); ++j)
      putHexInt8((sum[i] >> (j * 8)) & 0xff);
}
